src/main/java/com/nextcloud/common/NextcloudClient.kt

```diff
@@ -50,40 +47,47 @@ import javax.net.ssl.SSLSession
 import javax.net.ssl.TrustManager
 
 
-class NextcloudClient(var baseUri: Uri, val context: Context) {
+class NextcloudClient(var baseUri: Uri, val context: Context, client: OkHttpClient) {
     lateinit var credentials: String
     lateinit var userId: String
     lateinit var request: Request
     var followRedirects = true;
-    val client: OkHttpClient
+    val client = client
     
     companion object {
         @JvmStatic
         val TAG = NextcloudClient::class.java.simpleName
-    }
 
-    init {
-        val trustManager = AdvancedX509TrustManager(NetworkUtils.getKnownServersStore(context))
-        val sslContext = SSLContext.getInstance("TLSv1")
-        sslContext.init(null, arrayOf<TrustManager>(trustManager), null)
-        val sslSocketFactory = sslContext.socketFactory
+        private fun createDefaultClient(context: Context) : OkHttpClient {
+            val trustManager = AdvancedX509TrustManager(NetworkUtils.getKnownServersStore(context))
+            val sslContext = SSLContext.getInstance("TLSv1")
+            sslContext.init(null, arrayOf<TrustManager>(trustManager), null)
+            val sslSocketFactory = sslContext.socketFactory
 
-        client = OkHttpClient.Builder()
-                .cookieJar(CookieJar.NO_COOKIES)
-                .callTimeout(DEFAULT_DATA_TIMEOUT_LONG, TimeUnit.MILLISECONDS)
-                .sslSocketFactory(sslSocketFactory, trustManager)
-                .hostnameVerifier { asdf: String?, usdf: SSLSession? -> true }
-                .build()
+            return OkHttpClient.Builder()
+                    .cookieJar(CookieJar.NO_COOKIES)
+                    .callTimeout(DEFAULT_DATA_TIMEOUT_LONG, TimeUnit.MILLISECONDS)
+                    .sslSocketFactory(sslSocketFactory, trustManager)
+                    .hostnameVerifier { asdf: String?, usdf: SSLSession? -> true }
+                    .build()
+        }
     }
+
+    constructor(baseUri: Uri, context: Context) : this(baseUri, context, createDefaultClient(context))
    
     fun execute(remoteOperation: RemoteOperation): RemoteOperationResult {
-        return remoteOperation.run(this)
+        return try {
+            remoteOperation.run(this)
+        } catch (ex: Exception) {
+            RemoteOperationResult(ex)
+        }
     }
     
     fun execute(method: OkHttpMethodBase): Int {
         return method.execute(this)
     }
 
+    @Throws(IOException::class)
```

Comment:
@tobiasKaminsky This will proably break clients, but since `IOException` is checked, it must be handled.
There is no API to return valid `Response` in case of error, as `OkHttp3` use exceptions to signal abnormal condition. We should follow.

Reply:
Hm. I would go instead with try/catch so that we can return our own error code.
After all, this is in our library and also should only be called within the library.

Reply:
I wrapped the exception in a union-like object.